### PR TITLE
[PushNotificationsIOS] registration error event

### PR DESCRIPTION
**Pull request by nevir**
This is an updated version of #2336 and #7694.

---

This adds a `registrationError` event that is emitted by `PushNotificationIOS` whenever an application receives a registration error from APNS (APNS service failure, running on simulator, etc).  This event fires to the exclusion of the `register` event (and vice versa).

**How to use**

Add the following to your `AppDelegate.m`:

```obj-c
- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error
{
  [RCTPushNotificationManager didFailToRegisterForRemoteNotificationsWithError:error];
}
```

And register an event handler for the event:

```js
PushNotificationIOS.addEventListener('registrationError', function({ message, code }) {
  // Complete your registration process in error.
});
```

**Test plan**

Added support for this event (and `register`) to UIExplorer as a proof of concept.  Navigating to the push notifications example on a simulator is an easy way to reproduce this event.

### CODE REVIEW

**Comment by facebook-github-bot-4** [comment]
By analyzing the blame information on this pull request, we identified @nicklockwood and @JAStanton to be potential reviewers.

**Comment by facebook-github-bot-2** [comment]
@nevir updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9650/files/b727cc457b25900e46fb20965194a67620282b74..324f45e7c031b6c315c580349a402fdc29bf4a69)

**Comment by facebook-github-bot-1** [comment]
@nevir updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9650/files/324f45e7c031b6c315c580349a402fdc29bf4a69..fb0f5cee0f8d6c0f0d71e2d8194b8fe30fdf724c)

**Comment by facebook-github-bot-7** [comment]
@nevir updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9650/files/b727cc457b25900e46fb20965194a67620282b74..324f45e7c031b6c315c580349a402fdc29bf4a69)

File: Examples/UIExplorer/UIExplorer/AppDelegate.m [hunk 1]
```diff
@@ -76,4 +77,36 @@ - (void)loadSourceForBridge:(RCTBridge *)bridge
                             onComplete:loadCallback];
 }
 
+# pragma mark - Push Notifications
+
+// Required to register for notifications
+- (void)application:(UIApplication *)__unused application didRegisterUserNotificationSettings:(UIUserNotificationSettings *)notificationSettings
```
Comment (javache) [inline Examples/UIExplorer/UIExplorer/AppDelegate.m]: Please format like `application:(__unused UIApplication *)application didRegister...`

File: Examples/UIExplorer/js/PushNotificationIOSExample.js [hunk 2]
```diff
@@ -50,16 +50,18 @@ class Button extends React.Component {
 
 class NotificationExample extends React.Component {
   componentWillMount() {
-    // Add listener for push notifications
-    PushNotificationIOS.addEventListener('notification', this._onNotification);
-    // Add listener for local notifications
+    PushNotificationIOS.requestPermissions();
```
Comment (javache) [inline Examples/UIExplorer/js/PushNotificationIOSExample.js]: You should first add the listeners, and only then call `requestPermission`.

File: Libraries/PushNotificationIOS/PushNotificationIOS.js [hunk 3]
```diff
@@ -165,8 +167,8 @@ class PushNotificationIOS {
    */
   static addEventListener(type: string, handler: Function) {
```
Comment (javache) [inline Libraries/PushNotificationIOS/PushNotificationIOS.js]: Can you add a flowtype for this?

**Comment by javache** [comment]
Looks good, just a couple of small nits

**Comment by nevir** [comment]
@javache updated; thanks for the review!

**Comment by facebook-github-bot-2** [comment]
@nevir updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9650/files/fb0f5cee0f8d6c0f0d71e2d8194b8fe30fdf724c..7b658d521df3af7165a3d2821337a41bf7a7f2f5)

**Comment by javache** [comment]
@facebook-github-bot shipit

**Comment by facebook-github-bot-5** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3822142) to review internal test results.

**Pull request closed without merging**